### PR TITLE
Editor team

### PR DESCRIPTION
**Pull request by lxlxok**
the prototype edition for editor is ready for merge, we can test the editor by 3 step
step 1: 
http://127.0.0.1:8000/server/editor/create
step 2:
http://127.0.0.1:8000/server/editor/select
step 3:
http://127.0.0.1:8000/server/editor/test_edit

### CODE REVIEW

File: applications/server/controllers/editor.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    return dict(result = now)
+
+def select():
+    """
+    search for the exit procedure in the database
+    :return:
+    """
+    row= db(db.coding).select().first()
+    if row is not None:
+        rows = db(db.coding).select()
```
Comment (lucadealfaro) [inline applications/server/controllers/editor.py]: I don't understand what you are doing here.  Can you add some comments?  Why do you first get the first row, and then if there is one, loop over all the rows? 

Also, I would have imagined a controller that takes as input the device, and then displays what is the code for that device.

File: applications/server/controllers/editor.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        code_list = [{'id': r.id, 'procedures': r.procedures, 'times': r.times} for r in rows]
+        return dict(exit = 1, code_list = code_list)
```
Comment (lucadealfaro) [inline applications/server/controllers/editor.py]: Two things.  First, why are you doing a for loop if you then have a return in the loop? 
Second, I  think it's bad practice to return different dictionaries.  The web layout will have to work with all of them.  I think it's best here to have only one return statement, at the end.

File: applications/server/models/tables.py [hunk 3]
```diff
@@ -65,3 +65,17 @@
...
+
+
+
+
+
+
+
+
+## This is the table used to temperary testing editor
+db.define_table('coding',
```
Comment (lucadealfaro) [inline applications/server/models/tables.py:81]: Why don't you tie in this code a little bit more, adding a device_id field?

File: applications/server/controllers/editor.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                    editor_settings=preferences,
+                    id=code_id,
+                    data=data,
+                    )
+    plain_html = response.render('editor/edit_js.html', file_details)
+    file_details['plain_html'] = plain_html
+    return response.json(file_details)
+
+def save_procedure():
+    """
```
Comment (lucadealfaro) [inline applications/server/controllers/editor.py:42]: Can you please write comments that include the roles of all the parameters passed here? 
As a reader, I need to understand what this is doing precisely without reading the code.

File: applications/server/controllers/editor.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    return the save result for the saving procedure
+    """
+    code_id = request.vars.procedure_id
+    data = request.vars.procedure
+    stable = request.vars.stable
+    print(data)
```
Comment (lucadealfaro) [inline applications/server/controllers/editor.py]: Use logging, not printing.

File: applications/server/controllers/editor.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    #save the procedure data to the database
+    if stable is False:
+
+        #the final edition will use Team 2 API save(procedure_id, stable) to save the data
+        #we debug with temporary table to validate the function of editor
```
Comment (lucadealfaro) [inline applications/server/controllers/editor.py]: Are you talking with team2 about this?

File: applications/server/controllers/editor.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    This function is only used for self debug to show exited procedure in database
+    Team 2 provide the API get_procedures_for_user(user_id) to return the list of procedure_id belong to a user
+    This function will be deleted in the final edition
+    """
+    row= db(db.coding).select().first()
+    if row is not None:
+        rows = db(db.coding).select()
+        code_list = [{'id': r.id, 'procedures': r.procedures, 'times': r.times} for r in rows]
+        return dict(exit = 1, code_list = code_list)
+    return dict(exit = 0)
```
Comment (lucadealfaro) [inline applications/server/controllers/editor.py]: I would simply return code_list = None, so you don't have this spurious parameter exit.

Comment (lucadealfaro) [inline applications/server/controllers/editor.py]: In fact I think you can replace all of these lines simply with 

return dict(code_list=db(db.coding).select())

**Comment by lucadealfaro** [comment]
Can you take care of the few comments? 
Then you can do a quick demo in class today and we merge, ok?

File: applications/server/controllers/editor.py (hunk 6 above)
Comment (lxlxok) [inline applications/server/controllers/editor.py]: Yes, we have confirmed with team2 about this.

File: applications/server/controllers/editor.py (hunk 1 above)
Comment (lxlxok) [inline applications/server/controllers/editor.py]: This is used for self debuging and will be deleted finally.  When the table of procedure and database operation API have been completed by team2, we can use their API to select procedure and create procedure.

File: applications/server/controllers/editor.py (hunk 2 above)
Comment (lxlxok) [inline applications/server/controllers/editor.py]: This is only used for self debug and will be deleted finally. web layout team only need to integrated some code from test_editor in their web page. I have provided the encapsulated javascript function and HTML template for them.

File: applications/server/controllers/editor.py (hunk 5 above)
Comment (lxlxok) [inline applications/server/controllers/editor.py]: sorry, I forget to delete this debug code. I will use logging in the future.

File: applications/server/controllers/editor.py (hunk 4 above)
Comment (lxlxok) [inline applications/server/controllers/editor.py:42]: no problem , I will add the comment and push a new edition.  the editor_settings is the option parameters for editor feature. id is the procedure_id in the procedures TALBE, data is the code for procedure which is related with the id.
editor/edit_js.html will generated HTML code for editor with these parameters in file_details.

File: applications/server/controllers/editor.py (hunk 7 above)
Comment (lxlxok) [inline applications/server/controllers/editor.py]: No problem, I will change it and push it again.

**Comment by lxlxok** [comment]
Thanks so much for so many kindly advice. I will finished all the issues before the class today.

**Pull request merged**